api/agent/lb_agent.go

```diff
@@ -171,6 +188,17 @@ func (a *lbAgent) Submit(callI Call) error {
 
 	statsDequeueAndStart(ctx)
 
+	// pre-read and buffer request body if already not done based
+	// on GetBody presence.
+	buf, err := a.setRequestBody(ctx, call)
+	if err != nil {
+		logrus.WithError(err).Error("Failed to process call body")
+		return a.handleCallEnd(ctx, call, err, true)
+	}
+	if buf != nil {
```

Comment:
Does this block need to be swapped with the `err != nil {` block?  I ask because it looks like `setRequestBody` can return a buffer even if it returns an error.  Is the `bufPool.Put()` needed to free this back up?  If so, it _looks_ like you have the potential for a leak here.

Reply:
Arrgh. :-)